### PR TITLE
feat(awsutil-v2): implement awsutil for aws-sdk-go-v2

### PR DESCRIPTION
**Pull request by ddebko**
## Summary

This PR is adding a new major version of `awsutil`. The breaking change that requires a major version release is using the latest aws sdk: `aws-sdk-go-v2`. 

### Disclaimer:

Feel free to update to this major release change if you are migrating to `aws-sdk-go-v2`. The Vault team should be cautious of updating to this version due to potential breaking changes from behavior differences between the `aws-sdk-go` libraries.

### Note:

- This PR does not replicate the exact behaviors of the original v1 `awsutil` package. For example, environment variables are always loaded by default in v2. The aws-sdk-go-v2 also doesn't support `RemoteCredProviders`. GetRegion no longer can get a default region value from `ec2metadata` because it does not exist and the replacement `imds` pkg does not support the same feature.
- Custom endpoint resolvers are now associated with a client object (ex: iam, sts) rather than a aws.Config
- withStsEndpoint type was changed from a string to sts.EndpointResolverV2. This option flag was relabeled to withStsEndpointResolver.
- withIamEndpoint type was changed from a string to iam.EndpointResovlerV2. This option flag was relabeled to withIamEndpointResolver.

The benefits of changing the endpoint options from a string to their respective EndpointResolverV2 definition type is that now we can enable complex routing to a set of different endpoints based on fields defined in the resolverV2 struct. Please follow this [link](https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/endpoints/#endpointresolverv2-parameters) to read more about the EndpointResolverV2 option. Example:

```
type resolverV2 struct {
    // you could inject additional application context here as well
    region string
}

func (*resolverV2) ResolveEndpoint(ctx context.Context, params s3.EndpointParameters) (
        smithyendpoints.Endpoint, error,
    ) {
    if region == "A" {
        return smithyEndpoints.Endpoint{
            URI: url.Parse("https://custom.service.endpoint/"),
        }
    }
    if region == "B" {...}

    // delegate back to the default v2 resolver otherwise
    return s3.NewDefaultEndpointResolverV2().ResolveEndpoint(ctx, params)
}
```

### CODE REVIEW

**Comment by jefferai** [comment]
@ddebko rather than making this `awsutilv2` just make this `v2` of the `awsutil` package. That will make it a separate Go module that one must explicitly import and the current version can still be modified if desired. But the general line of development is clearly to catch up with the current, supported version of the AWS SDK.

**Comment by sgmiller** [comment]
Jeff's suggestion might make it easier to review too.

**Comment by ddebko** [comment]
@jefferai @sgmiller I have updated the PR to move the code into a subdirectory/package called v2 in awsutil. I have 2 `TODO` comments for the custom endpoint resolvers and I was hoping to get some feedback on the behavior of the resolvers.

**Comment by jefferai** [comment]
Just a note to please move the files back to the main directory and update `go.mod` before tagging as v2.

Arguably it'd be more useful to have them there now as we could see the differences instead of it all being new code.

File: awsutil/v2/generate_credentials.go
Comment (ddebko) [reply awsutil/v2/generate_credentials.go:260]: If both the AccessKey and SecretKey are empty strings, then the conditional will return false. This conditional check is here to validate that both AccessKey and SecretKey are present. I think that AWS returns a generic credential error message if you don't supply the AccessKey, but provide the SecretKey (and vice-versa). By adding this conditional check, we are providing the user more context into the issue.

**Review by sgmiller — APPROVED** [review]
Afaict it looks good to me.

File: awsutil/v2/generate_credentials.go
Comment (hugoamvieira) [reply awsutil/v2/generate_credentials.go]: Sorry, I misread the `if` statement 🤦

**Review by hugoamvieira — APPROVED** [review]

**Review by jefferai — CHANGES_REQUESTED** [review]
Please make sure to make the necessary changes to `go.mod` given the version bump.

Unless the plan is to mark this as v1.0.0 since the current tags are still in `v0`?

**Comment by ddebko** [comment]
@jefferai what changes do you expect to see in the go.mod file?
I thought we were going to version bump based on release tags as we do with the other repos. Does this still require a bump in the go.mod? Is this the change you would like to see added?
```
module github.com/hashicorp/go-secure-stdlib/awsutil/v1
```

**Comment by ddebko** [comment]
@jefferai what changes do you expect to see in the go.mod file?
I thought we were going to version bump based on release tags as we do with the other repos. Does this still require a bump in the go.mod? Is this the change you would like to see added?
```
module github.com/hashicorp/go-secure-stdlib/awsutil/v1
```

**Comment by ddebko** [comment]
> Please make sure to make the necessary changes to `go.mod` given the version bump.
> 
> Unless the plan is to mark this as v1.0.0 since the current tags are still in `v0`?

I think we should leave the go.mod as is because the next major version bump should be v1.0.0

File: awsutil/README.md
Comment (austingebauer) [inline awsutil/README.md:31]: Thanks!

File: awsutil/README.md
Comment (austingebauer) [inline awsutil/README.md:33]: Mind adding a link to the branch? Can go around the `awsutil/v0` text. Doesn't look like the branch exists yet.

Comment (ddebko) [reply awsutil/README.md]: I created the branch and added a link to this line.

**Review by jefferai — APPROVED** [review]
Approving to unblock merge access

**Comment by benashz** [comment]
For some reason the changes from 1a4b95565d57828244cbb7879b8d3a8f3d20522a were not ported over. The PR description does not seem to mention why they were dropped.

**Pull request merged**
